File: src/pose_estimator.cpp
//
// Created by noah on 11/12/17.
//

#include "pose_estimator.h"

/**
 * Default constructor.
 *
 * @param objectCorners The absolute, 3D position of the corners of the actual object.
 * @param cameraMatrix The matrix describing how the camera distorts images.
 * @param distortionCoefficients The matrix of coefficients for how the camera is distorted.
 */
pose_estimator::pose_estimator(const vector<Point3f> &objectCorners, const Mat1f &cameraMatrix,
                               const Mat1f &distortionCoefficients) {
	this->objectCorners = objectCorners;
	this->cameraMatrix = cameraMatrix;
	this->distortionCoefficients = distortionCoefficients;
}

/**
 * Get the position and rotation of the camera.
 * @param corners The 4 corners of the object as detected by the camera.
 * @return A 4x4 matrix with top-left 3x3 describing the rotation of the camera and the right column giving x,y,z position.
 */
Mat1f pose_estimator::get_pose(const vector<Point2f> &corners) {
	rvec = Mat();
	tvec = Mat();
	solvePnP(objectCorners, corners, cameraMatrix, distortionCoefficients, rvec, tvec);
	Rodrigues(rvec, zyx);
	hconcat(zyx, tvec, final);
	vconcat(final,bottom, final);
	final = final.inv();
	return final;
}

File: src/pose_estimator.py
import cv2 as cv
import numpy as np


class PoseEstimator:

    def __init__(self, object_corners, camera_matrix, dist_coeffs):
        self.object_corners = np.ndarray.astype(object_corners, np.float32)
        self.camera_matrix = np.ndarray.astype(camera_matrix, np.float32)
        self.dist_coeffs = np.ndarray.astype(dist_coeffs, np.float32)
        self.bottom = np.array([[0, 0, 0, 1]], dtype=np.float32)
        self.zyx = np.array([0])
        self.final = np.array([0])

    def get_pose(self, corners):
        # _, rvec, tvec = cv.solvePnP(self.object_corners, np.array(corners), self.camera_matrix, self.dist_coeffs, flags=cv.SOLVEPNP_P3P)
        _, rvec, tvec = cv.solvePnP(self.object_corners, np.array(corners), self.camera_matrix, self.dist_coeffs)
        zyx, _ = cv.Rodrigues(rvec)
        final = np.concatenate((zyx, tvec), axis=1)
        final = np.concatenate((final, self.bottom), axis=0)
        return np.linalg.inv(final)


